platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java

```diff
@@ -48,14 +67,25 @@ public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
     }
 
     @VisibleForTesting
-    static void buildProblem(Diagnostic<? extends JavaFileObject> diagnostic, ProblemSpec spec) {
+    void buildProblem(Diagnostic<? extends JavaFileObject> diagnostic, ProblemSpec spec) {
         spec.id(mapKindToId(diagnostic.getKind()), mapKindToLabel(diagnostic.getKind()), GradleCoreProblemGroup.compilation().java());
         spec.severity(mapKindToSeverity(diagnostic.getKind()));
-        addDetails(spec, diagnostic.getMessage(Locale.getDefault()));
+        addFormattedMessage(spec, diagnostic);
```

Comment:
The ID is a bit tricky. What do we consider and ID here. My current thinking that we should have different IDs for different compilation problems. For example, an unused variable woudld have a group of `compilation:java` and an id called `unused-variable`. We don't have to do it now; this is already a big enough change, but another thing that we should track. Please create an issue for that.

Reply:
The `JCDiagnostics` class still has a few tricks up its sleeve (e.g., `JCDiagnostic#getCode()`), which could allow it to cut up the compiler's problem space much finer.
I suggest we keep that issue for a later PR; added it to our further enhancements list.